StreamChatClient_v3/Controllers/EntityDatabaseObserver.swift

```diff
@@ -0,0 +1,119 @@
+            self = .remove(item)
+        }
+    }
+}
+
+/// Observes changes of a single entity specified using an `NSFetchRequest`in the provided `NSManagedObjectContext`.
+class EntityDatabaseObserver<Item, DTO: NSManagedObject> {
+    /// The observed item. `nil` of no item matches the predicate or the item was deleted.
+    @Cached var item: Item?
+    
+    /// Called with the aggregated changes after the internal `NSFetchResultsController` calls `controllerDidChangeContent`
+    /// on its delegate.
+    var onChange: ((EntityChange<Item>) -> Void)? {
+        didSet {
+            changeAggregator.onChange = { [unowned self] listChanges in
+                log.assert(listChanges.count <= 1, "EntityDatabaseObserver predicate shouldn't produce more than one change")
+                if let entityChange = listChanges.first.map(EntityChange.init) {
+                    self._item.reset()
+                    self.onChange?(entityChange)
+                }
+            }
+        }
+    }
+    
+    /// Acts like the `NSFetchedResultsController`'s delegate and aggregates the reported changes into easily consumable form.
+    private(set) lazy var changeAggregator: ListChangeAggregator<DTO, Item> =
+        ListChangeAggregator<DTO, Item>(itemCreator: self.itemCreator)
+    
+    /// Used for observing the changes in the DB.
+    private(set) lazy var frc: NSFetchedResultsController<DTO> = self.fetchedResultsControllerType
+        .init(fetchRequest: self.request,
+              managedObjectContext: self.context,
+              sectionNameKeyPath: nil,
+              cacheName: nil)
+    
+    /// The `NSFetchedResultsController` subclass the observe uses to create its FRC. You can inject your custom subclass
+    /// here if needed, i.e. when testing.
+    var fetchedResultsControllerType: NSFetchedResultsController<DTO>.Type = NSFetchedResultsController<DTO>.self
+    
+    let itemCreator: (DTO) -> Item?
+    let request: NSFetchRequest<DTO>
+    let context: NSManagedObjectContext
+    
+    /// Creates a new `ListObserver`.
+    ///
+    /// Please note that no updates are reported until you call `startUpdating`.
+    ///
+    /// - Important: ⚠️ Because the observer uses `NSFetchedResultsController` to observe the entity in the DB, it's required
+    /// that the provided `fetchRequest` has at lease one `NSSortDescriptor` specified.
+    ///
+    /// - Parameters:
+    ///   - context: The `NSManagedObjectContext` the observer observes.
+    ///   - fetchRequest: The `NSFetchRequest` that specifies the elements of the list.
+    ///   - itemCreator: A close the observe uses to convert DTO objects into Model objects.
+    init(context: NSManagedObjectContext,
+         fetchRequest: NSFetchRequest<DTO>,
+         itemCreator: @escaping (DTO) -> Item?) {
+        self.context = context
+        request = fetchRequest
+        self.itemCreator = itemCreator
+        
+        _item.computeValue = { [unowned self] in
```

Comment:
I can't find the definition of `_item` anywhere 😅 Is this a feature of property wrappers?

Reply:
Yes. `item` is the wrapped value, `_item` is the property wrapper itself.

Reply:
ooooh TIL